Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,249 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SharedAudioDestination.h"
+
+#include "AudioUtilities.h"
+#include <wtf/WTFSemaphore.h>
+#include <wtf/WeakPtr.h>
+#include <wtf/WorkQueue.h>
+
+namespace WebCore {
+
+class SharedAudioDestinationAdapter : public RefCounted<SharedAudioDestinationAdapter>, public CanMakeWeakPtr<SharedAudioDestinationAdapter>, public AudioIOCallback {
+public:
+    using AudioDestinationCreationFunction = SharedAudioDestination::AudioDestinationCreationFunction;
+    static Ref<SharedAudioDestinationAdapter> ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction);
+
+    void addRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+    void removeRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+
+    unsigned framesPerBuffer() const
+    {
+        return m_workBus->length();
+    }
+
+private:
+    using AdapterKey = std::tuple<unsigned, float>;
+    using AdapterMap = HashMap<AdapterKey, WeakPtr<SharedAudioDestinationAdapter>>;
+    static AdapterMap& sharedMap();
+
+    SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, AudioDestinationCreationFunction&&);
+
+    void render(AudioBus* sourceBus, AudioBus* destinationBus, size_t framesToProcess, const AudioIOPosition& outputPosition) final;
+    void isPlayingDidChange() final { }
+
+    void configureRenderThread(CompletionHandler<void(bool)>&&);
+
+    Ref<AudioDestination> m_destination;
+    Ref<AudioBus> m_workBus;
+    Ref<WorkQueue> m_configurationQueue;
+
+    Lock m_renderLock;
+    Semaphore m_configurationSemaphore;
+
+    // Only access on MainThread
+    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+    bool m_started { false };
+
+    // Protected by m_renderLock:
+    bool m_needsConfiguration { true };
+    Vector<RefPtr<SharedAudioDestination>> m_newRenderers;
+    Vector<RefPtr<SharedAudioDestination>> m_configuredRenderers;
+};
+
+auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+{
+    static NeverDestroyed<AdapterMap> map;
```

Comment:
Is this always accessed on the main thread? I assume so since I see no locks.
If so, please use `MainThreadNeverDestroyed`.